File: Data_Structures/Assignment 5/sorting.py
# sorting.py CS2028 2016 Cheng
# Usage: python sorting.py
# time spent for each sorting algorithm on identical randomized sequences of 
# increasing sizes

import random
import time, datetime

# used by selectionSort
# returns the index of the smallest element in seq starting at start
def select(seq, start):
	minIndex = start
	for i in range(start+1, len(seq)):
		if seq[minIndex] > seq[i]:
			minIndex = i
	return minIndex

def selectionSort(seq):
	for i in range(len(seq)-1):
		minIndex = select(seq, i)
		tmp = seq[i]
		seq[i] = seq[minIndex]
		seq[minIndex] = tmp

# instead returns None when key is not in seq, it returns the index it should be at
def binarysearch(seq, high, key):
	lo = 0
	hi = high
	while lo <= hi:
		mid = (lo + hi) // 2
		if seq[mid] == key:
			return mid
		elif key < seq[mid]:
			hi = mid - 1
		else:
			lo = mid + 1
	return lo

# used by insertionSort
# precondition: seq[0:index] is sorted
# postcondition: seq[0:index + 1] is sorted
def insert(seq, index):
	value = seq[index]
	i = index -1
	while i >= 0 and seq[i] > value:
		seq[i + 1] = seq[i]
		i -= 1
	seq[i + 1] = value

# used by insertionSortWithBinarysearch
# precondition: seq[0:index] is sorted
# postcondition: seq[0:index + 1] is sorted
# binary search used on sorted seq[0:index] 
# faster than insert without binary search
def insertWithBinarysearch(seq, index):
	value = seq[index]
	lo = binarysearch(seq, index, value)
	for i in range(index - 1, lo - 1, -1):
		seq[i + 1] = seq[i]
	seq[lo] = value

def insertionSort(seq):
	for i in range(1, len(seq)):
		insert(seq, i)

def insertionSortWithBinarysearch(seq):
	for i in range(1, len(seq)):
		insertWithBinarysearch(seq, i)

# When seq is correctly sorted its ith element shuld be i.
def checkSorting(seq):
	for i in range(len(seq)):
		if seq[i] != i:
			print("error")

# precondition: seq[start:mid] and seq[mid:stop] are sorted
# postconditio: seq[start:stop] is sorted
def merge(seq, start, mid, stop):
	lst = []
	i = start
	j = mid
	while i < mid and j < stop:
		if seq[i] < seq[j]:
			lst.append(seq[i])
			i+=1
		else:
			lst.append(seq[j])
			j+=1
	while i < mid:
		lst.append(seq[i])
		i+=1
	for i in range(len(lst)):
		seq[start + i] = lst[i]

def mergeSortRecursively(seq, start, stop):
	if start >= stop - 1:
		return
	mid = (start + stop) // 2
	mergeSortRecursively(seq, start, mid)
	mergeSortRecursively(seq, mid, stop)
	merge(seq, start, mid, stop)

def mergeSort(seq):
	mergeSortRecursively(seq, 0, len(seq))

# postcondition: returns j so that all elements in seq[start:j] are smaller than seq[j] and 
#        all elements in seq[j+1:stop] are not smaller than seq[j]
def partition(seq, start, stop):
	pivotIndex = start
	pivot = seq[pivotIndex]
	i = start+1
	j = stop-1
	while i <= j:
		while i <= j and not pivot < seq[i]:
			i+=1
		while i <= j and pivot < seq[j]:
			j-=1
		if i < j:
			tmp = seq[i]
			seq[i] = seq[j]
			seq[j] = tmp
			i+=1
			j-=1
	seq[pivotIndex] = seq[j]
	seq[j] = pivot
	return j

# random shuffle to rearrange elements of seq
def randomize(seq):
	for j in range(1, len(seq)):
		tmp = seq[j]
		pos = random.randint(0, j - 1)
		seq[j] = seq[pos]
		seq[pos] = tmp

def quicksortRecursively(seq, start, stop):
	if start >= stop:
		return 
	pivotIndex = partition(seq, start, stop)
	quicksortRecursively(seq, start, pivotIndex)
	quicksortRecursively(seq, pivotIndex+1, stop)

def quicksort(seq):
	randomize(seq)
	quicksortRecursively(seq, 0, len(seq))

# postcondition: the largest element at seq[stop - 1]
# returns True when seq has been mutated and False otherwise
def bubbleRound(seq, stop):
	bubbled = False
	for i in range(1, stop):
		if seq[i] < seq[i - 1]:
			tmp = seq[i]
			seq[i] = seq[i - 1]
			seq[i - 1] = tmp
			bubbled = True
	return bubbled

def bubbleSort(seq):
	stop = len(seq)
	while bubbleRound(seq, stop):
		stop-=1

# using the built-in sort function of Python list
def builtinSort(seq):
	seq.sort()

# apply six sorting algorithms on the same randomized seq and report times
# repeat this on ever larger seq (1000, 2000, ..., 6000)
def compareAlgorithms():

	# seq will not be mutated
	# lst is a clone of seq and will be sorted using algorithm
	# time in miliseconds will be returned as time spent
	def sortWithAlgorithm(seq, algorithm):
		size = len(seq)
		lst2 = [None] * size
		for j in range(size):
			lst2[j] = seq[j]
		starttime = datetime.datetime.now()
		algorithm(lst2)
		endtime = datetime.datetime.now()
		checkSorting(lst2)
		return (endtime - starttime).total_seconds() * 1000

	for N in range(6):
		size = 1000 * (N + 1)
		lst = [None] * size
		for j in range(size):
			lst[j] = j
		randomize(lst)
		S = sortWithAlgorithm(lst, selectionSort)
		I = sortWithAlgorithm(lst, insertionSort)
		IB = sortWithAlgorithm(lst, insertionSortWithBinarysearch)
		M = sortWithAlgorithm(lst, mergeSort)
		Q = sortWithAlgorithm(lst, quicksort)
		B = sortWithAlgorithm(lst, bubbleSort)
		P = sortWithAlgorithm(lst, builtinSort)
		print(S, I, IB, M, Q, B, P)

def main():
	compareAlgorithms()	

if __name__ == "__main__":
	main()

File: Data_Structures/Assignment 5/sorting.cpp
/* sorting.cpp CS2028 2016 Cheng
   implementing sorting.py in C++ with arrays
   Usage: g++ sorting.cpp
   Usage: ./a
*/
#include <stdio.h>
#include <stdlib.h>
#include <time.h>

int select(int *seq, int seqLen, int start){
    int minIndex = start;
    for (int i = start + 1; i < seqLen; i++)
        if (seq[minIndex] > seq[i]) minIndex = i;
    return minIndex;
}

void selectionSort(int *seq, int seqLen){
    for (int i = 0; i < seqLen - 1; i++){
        int minIndex = select(seq, seqLen, i);
        int tmp = seq[i];
        seq[i] = seq[minIndex];
        seq[minIndex] = tmp;
    }
}

void randomize(int *seq, int seqLen){
    for (int j = 1; j < seqLen; j++){
        int tmp = seq[j];
        int pos = rand() % j;
        seq[j] = seq[pos];
        seq[pos] = tmp;
    }
}

void checkSorting(int *seq, int seqLen){
    for (int i = 0; i < seqLen; i++) if (seq[i] != i) printf("error\n");
}

int binarysearch(int *seq, int high, int key){
    int lo = 0;
    int hi = high;

    // Check if key is equal to, less than, or greater than mid.
    // If it is equal, return mid index. If it is less than,
    // resume search in lower half of sequence. If it is greater
    // than, resume search in upper half of sequence.
    while (lo <= hi) {
        int mid = (lo + hi) / 2;

        if (seq[mid] == key) {
            return mid;
        } else if (key < seq[mid]) {
            hi = mid - 1;
        } else {
            lo = mid + 1;
        }
    }

    return lo;
}

void insert(int *seq, int index){
    int value = seq[index];
    int i = index - 1;

    // Start iteration through sequence backwards
    // and keep iterating until a value greater than
    // value at index is found.

    while (i >= 0 && seq[i] > value) {
        seq[i + 1] = seq[i];
        i--;
    }

    seq[i + 1] = value;
}

void insertWithBinarysearch(int *seq, int index){
    int value = seq[index];
    int lo = binarysearch(seq, index, value);
    for (int i = index - 1; i > lo - 1; i--) {
        seq[i + 1] = seq[i];
    }

    seq[lo] = value;
}

void insertionSort(int *seq, int seqlen){
    for (int i = 1; i < seqlen; i++) insert(seq, i);
}

void insertionSortWithBinarysearch(int *seq, int seqlen){
    for (int i = 1; i < seqlen; i++) insertWithBinarysearch(seq, i);
}

int sortWithAlgorithm(int *seq, int seqLen, void (*algorithm)(int*, int)){
    clock_t starttime, endtime;
    int lst2[seqLen];
    for (int j = 0; j < seqLen; j++) lst2[j] = seq[j];
    starttime = clock();
    algorithm(lst2, seqLen);
    endtime = clock();
    checkSorting(lst2, seqLen);
    return endtime - starttime;
}

void merge(int *seq, int start, int mid, int stop){
    int lst[stop - start];
    int i = start;
    int j = mid;
        int pos = 0;
    while (i < mid && j < stop)
        if (seq[i] < seq[j]) lst[pos++] = seq[i++];
        else lst[pos++] = seq[j++];
    while (i < mid) lst[pos++] = seq[i++];
    for (i = 0; i < pos; i++) seq[start + i] = lst[i];
}

void mergeSortRecursively(int *seq, int start, int stop){
    if (start >= stop - 1) {
        return;
    }

    // Make recursive calls until seq becomes only one item.
    int mid = (start + stop) / 2;
    mergeSortRecursively(seq, start, mid);
    mergeSortRecursively(seq, mid, stop);
    merge(seq, start, mid, stop);
}

void mergeSort(int *seq, int seqlen){
    mergeSortRecursively(seq, 0, seqlen);
}

int partition(int *seq, int start, int stop){
    int pivotIndex = start;
    int pivot = seq[pivotIndex];
    int i = start + 1;
    int j = stop - 1;

    // Loop through list, put all values less than pivot value
    // on left side of pivot, put all values greater than pivot
    // on right side of pivot.
    while (i <= j) {
        while (i <= j && pivot >= seq[i]) {
            i++;
        }

        while (i <= j && pivot < seq[j]) {
            j--;
        }

        if (i < j) {
            int temp = seq[i];
            seq[i] = seq[j];
            seq[j] = temp;
            i++;
            j--;
        }
    }

    seq[pivotIndex] = seq[j];
    seq[j] = pivot;
    return j;
}

void quicksortRecursively(int *seq, int start, int stop){
    if (start >= stop) {
        return;
    }

    // Keep making quicksort recursive call until there is only
    // one item in sequence.
    int pivotIndex = partition(seq, start, stop);
    quicksortRecursively(seq, start, pivotIndex);
    quicksortRecursively(seq, pivotIndex + 1, stop);
}

void quicksort(int *seq, int seqlen){
    randomize(seq, seqlen);
    quicksortRecursively(seq, 0, seqlen);
}

bool bubbleRound(int *seq, int stop){
    bool bubbled = false;

    for (int i = 1; i < stop; i++) {
        if (seq[i] < seq[i - 1]) {
            int tmp = seq[i];
            seq[i] = seq[i - 1];
            seq[i - 1] = tmp;
            bubbled = true;
        }
    }

    return bubbled;
}

void bubbleSort(int *seq, int seqlen){
    int stop = seqlen;
    while (bubbleRound(seq, stop--));
}

void compareAlgorithms(){
    for (int N = 0; N < 6; N++){
        int size = 10000 * (N + 1);  // ten times longer than Python
        int lst[size];
        for (int j = 0; j < size; j++) lst[j] = j;
        randomize(lst, size);
        int S = sortWithAlgorithm(lst, size, selectionSort);
        int I = sortWithAlgorithm(lst, size, insertionSort);
        int IB = sortWithAlgorithm(lst, size, insertionSortWithBinarysearch);
        int M = sortWithAlgorithm(lst, size, mergeSort);
        int Q = sortWithAlgorithm(lst, size, quicksort);
        int B = sortWithAlgorithm(lst, size, bubbleSort);
        printf("%d %d %d %d %d %d\n", S, I, IB, M, Q, B);
    }
}

int main(){
    compareAlgorithms();
}
